File: lib/cookies.js
const fs = require("fs");
const path = require("path");
const crypto = require("crypto");

const COOKIE_DIR = process.env.COOKIE_DIR || "./cookies";

function cookiePath(token) {
  // Hash rather than slice the raw token — two tokens sharing the same
  // first N characters (plausible given common token-format prefixes)
  // would otherwise silently overwrite each other's saved cookies.
  const hash = crypto.createHash("sha256").update(token).digest("hex").slice(0, 16);
  return path.join(COOKIE_DIR, `${hash}.json`);
}

function loadCookies(token) {
  try {
    const p = cookiePath(token);
    if (fs.existsSync(p)) return JSON.parse(fs.readFileSync(p, "utf-8"));
  } catch (_) {}
  return null;
}

function saveCookies(token, cookies) {
  try {
    if (!fs.existsSync(COOKIE_DIR)) fs.mkdirSync(COOKIE_DIR, { recursive: true });
    fs.writeFileSync(cookiePath(token), JSON.stringify(cookies, null, 2));
  } catch (err) {
    console.error(`[cookies] Failed to save: ${err.message}`);
  }
}

function clearCookies(token) {
  try {
    const p = cookiePath(token);
    if (fs.existsSync(p)) fs.unlinkSync(p);
  } catch (_) {}
}

module.exports = { cookiePath, loadCookies, saveCookies, clearCookies };

File: lib/cookies.py
"""Port of lib/cookies.js — per-token cookie persistence, keyed by token hash
(not a raw prefix) so two tokens sharing a common prefix never collide."""
import hashlib
import json
import os
from pathlib import Path
from typing import Optional

COOKIE_DIR = Path(os.environ.get("COOKIE_DIR", "./cookies"))


def cookie_path(token: str) -> Path:
    token_hash = hashlib.sha256(token.encode()).hexdigest()[:16]
    return COOKIE_DIR / f"{token_hash}.json"


def load_cookies(token: str) -> Optional[list]:
    try:
        p = cookie_path(token)
        if p.exists():
            return json.loads(p.read_text(encoding="utf-8"))
    except Exception:
        pass
    return None


def save_cookies(token: str, cookies: list) -> None:
    try:
        COOKIE_DIR.mkdir(parents=True, exist_ok=True)
        cookie_path(token).write_text(json.dumps(cookies, indent=2), encoding="utf-8")
    except Exception as e:
        print(f"[cookies] Failed to save: {e}")


def clear_cookies(token: str) -> None:
    try:
        p = cookie_path(token)
        if p.exists():
            p.unlink()
    except Exception:
        pass
